### ISSUE TITLE
Pre-train model for Javascript

### ISSUE THREAD

**Issue by Alfusainey** [tuvistavie/bigcode-tools#9]
@tuvistavie is there an existing pre-trained model for Javascript that one can use out-of-the-box? i have three data sets for training, testing and validation and I want to generated emebeddings for each dataset.

**Comment by tuvistavie** [comment]
Unfortunately I do not have pretrained embeddings for JavaScript.
Note that you should normally train a single embedding on a large corpus and use it for training, test and validation, rather than train a different embedding per dataset.

You can either use your data directly, use the [fetcher module](https://github.com/tuvistavie/bigcode-tools/tree/master/bigcode-fetcher) to collect some data to generate the embedding, or use an existing dataset such as the one available here:
http://learnbigcode.github.io/datasets/
This tool should be compatible with the format of the ASTs in the above link.

Please let me know if you run into any trouble.

**Comment by Alfusainey** [comment]
@tuvistavie thank you. i am using this dataset http://www.srl.inf.ethz.ch/js150.php.

i am even beginning to wonder if I really need an embedding now. am working on a pet project to provide support for auto-completion (using a vanilla LSTM model) just for fun. i was thinking that the input to the LSTM model should be a learned embedding (instead of a token-sequence). this is why I wanted to create an embedding for each data set before passing it to the LSTM.

what do you think: does it make sense to use the generated skipgram like and use that as input to the LSTM model? Or should I generate a vocabulary (with application label tokens) and feed that to the LSTM model? 

I am just wondering how the input to the LSTM model should look like and would appreciate your suggestion

**Comment by tuvistavie** [comment]
What you are trying to do makes sense, but the skipgram usage is I think a little off.
This should rather be a two steps process:

1. Train embeddings
code -> asts -> skipgram-data -> embeddings

2. Train the LSTM
code -> asts -> embeddings learned in 1 -> LSTM

You can choose to fine-tune the embeddings when training your LSTM, but you should normally only train the embeddings once.

Does this make sense?

**Comment by Alfusainey** [comment]
Thank you @tuvistavie. yes it makes sense. i was wondering if I should use all my data (train + test + val) when training the embedding?

in step 2: should I use the JSON serialization of the AST together with the learned embeddings as input to the LSTM? i.e code **->** (asts, learned embeddings) **->** LSTM. am just a bit confuse moving from asts -> embeddings learned in 1 (since we already have the embeddings from step 1)

**Comment by tuvistavie** [comment]
Sorry for the delay.

> i was wondering if I should use all my data

I suggest you use only your training data when learning the embeddings,
to be sure what you learned generalizes well enough.

> in step 2: should I use the JSON serialization of the AST together with the learned embeddings as input to the LSTM?

Yes, you should use the vocabulary and embeddings from step 1.

**Comment by buaanlsdewm** [comment]
@tuvistavie hello,i use your bigcode-tools to generate token embeddings.i get problems when running the code ‘docker-bigcode bigcode-ast-tools generate-vocabulary --strip-identifiers -o workspace/java-vocabulary-no-ids.tsv workspace/apache-commons-asts.json’ in step 4 in the tutorial，you can see the bug in the following image.Can you solve the problem?thank you
![1](https://user-images.githubusercontent.com/38215243/66464390-78e69880-eab1-11e9-8808-567854829a87.PNG)
